Q1. Handle Missing Values

In [2]:
import pandas as pd

df = pd.read_csv("employee_productivity_dataset.csv")

df['Age'].fillna(df['Age'].median(), inplace=True)
df['Salary'].fillna(df['Salary'].mean(), inplace=True)
df['Hours_Worked_Per_Week'].fillna(df['Hours_Worked_Per_Week'].median(), inplace=True)
df['Performance_Score'].fillna(df['Performance_Score'].mean(), inplace=True)

print(df.head())

  Employee_ID         Name Department   Age  Gender Work_Mode   Join_Date  \
0        E001  Amit Sharma         IT  25.0    Male    Remote  15-01-2021   
1        E002   Riya Patel         HR  30.0  Female    Hybrid  15-02-2020   
2        E003  John Mathew    Finance  29.0    Male    Onsite  10-03-2022   
3        E004  Sneha Joshi         IT  28.0  Female    Remote  22-04-2021   
4        E005  Arjun Verma  Marketing  35.0    Male    Hybrid  18-05-2020   

   Hours_Worked_Per_Week  Projects_Completed  Performance_Score        Salary  \
0                   40.0                   5             4.5000  50000.000000   
1                   38.0                   4             4.0000  45000.000000   
2                   45.0                   6             4.3125  60000.000000   
3                   41.0                   5             4.2000  52000.000000   
4                   50.0                   7             4.8000  50666.666667   

    Location  
0       Pune  
1     Mumbai  
2  Ba

/tmp/ipykernel_3031/390858693.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Age'].fillna(df['Age'].median(), inplace=True)
/tmp/ipykernel_3031/390858693.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try 

Q2. Label Encoding

In [3]:
from sklearn.preprocessing import LabelEncoder

le_gender = LabelEncoder()
le_department = LabelEncoder()

df['Gender'] = le_gender.fit_transform(df['Gender'])
df['Department'] = le_department.fit_transform(df['Department'])

print(df[['Gender','Department']].head())

   Gender  Department
0       1           2
1       0           1
2       1           0
3       0           2
4       1           3


Q3. One-Hot Encoding

In [4]:
df = pd.get_dummies(df,
                    columns=['Work_Mode','Location'])

print(df.head())
print("Total Columns:", len(df.columns))

  Employee_ID         Name  Department   Age  Gender   Join_Date  \
0        E001  Amit Sharma           2  25.0       1  15-01-2021   
1        E002   Riya Patel           1  30.0       0  15-02-2020   
2        E003  John Mathew           0  29.0       1  10-03-2022   
3        E004  Sneha Joshi           2  28.0       0  22-04-2021   
4        E005  Arjun Verma           3  35.0       1  18-05-2020   

   Hours_Worked_Per_Week  Projects_Completed  Performance_Score        Salary  \
0                   40.0                   5             4.5000  50000.000000   
1                   38.0                   4             4.0000  45000.000000   
2                   45.0                   6             4.3125  60000.000000   
3                   41.0                   5             4.2000  52000.000000   
4                   50.0                   7             4.8000  50666.666667   

   Work_Mode_Hybrid  Work_Mode_Onsite  Work_Mode_Remote  Location_Bangalore  \
0             False      

Q4. Normalization (Min-Max Scaling)

In [5]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df[['Salary','Hours_Worked_Per_Week']] = scaler.fit_transform(
    df[['Salary','Hours_Worked_Per_Week']]
)

print(df[['Salary','Hours_Worked_Per_Week']].head())

     Salary  Hours_Worked_Per_Week
0  0.347826               0.285714
1  0.130435               0.142857
2  0.782609               0.642857
3  0.434783               0.357143
4  0.376812               1.000000


Q5. Standardization (Scaling)

In [6]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

df[['Age','Projects_Completed']] = scaler.fit_transform(
    df[['Age','Projects_Completed']]
)

print(df[['Age','Projects_Completed']].head())

        Age  Projects_Completed
0 -1.172018           -0.211604
1  0.000000           -0.916949
2 -0.234404            0.493742
3 -0.468807           -0.211604
4  1.172018            1.199088


Q6. Compare Scaling Methods

In [7]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler

comparison = pd.DataFrame()

comparison['Original_Salary'] = df['Salary']

comparison['MinMax_Salary'] = MinMaxScaler().fit_transform(
    df[['Salary']]
)

comparison['Standardized_Salary'] = StandardScaler().fit_transform(
    df[['Salary']]
)

print(comparison.head())

   Original_Salary  MinMax_Salary  Standardized_Salary
0         0.347826       0.347826        -9.745089e-02
1         0.130435       0.130435        -8.283326e-01
2         0.782609       0.782609         1.364312e+00
3         0.434783       0.434783         1.949018e-01
4         0.376812       0.376812        -3.732632e-16


Q7. Build Preprocessing Pipeline

In [13]:
numeric_features = [
    'Age',
    'Hours_Worked_Per_Week',
    'Projects_Completed',
    'Performance_Score',
    'Salary'
]

categorical_features = [
    'Gender',
    'Department'
]

preprocessor = ColumnTransformer(
    [
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'),
         categorical_features)
    ]
)

pipeline = Pipeline([
    ('preprocessor', preprocessor)
])

print("Pipeline:", pipeline)

Pipeline: Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', StandardScaler(),
                                                  ['Age',
                                                   'Hours_Worked_Per_Week',
                                                   'Projects_Completed',
                                                   'Performance_Score',
                                                   'Salary']),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['Gender', 'Department'])]))])


Q8. Apply Pipeline

In [14]:
X = pipeline.fit_transform(df)

print("Final Shape:", X.shape)

Final Shape: (250, 11)


Q9. Conceptual Question

*   Why is scaling important in Python?



# ANS:

Scaling is important because many Machine Learning algorithms use distance calculations. Features with large values (e.g., Salary = 50,000) can dominate features with small values (e.g., Performance Score = 4.5). Scaling puts all features on a similar range, improving:



*   Model accuracy,Faster convergence
*   Better performance of algorithms such as KNN, SVM, Linear Regression,
  and Neural Networks



Q10. Conceptual Question


*   Why do we convert categorical data into numerical form?



# ANS:
Machine Learning algorithms work with mathematical computations and cannot directly understand text values.

**Gender:**

*   Male
*   Female

*After Encoding*

Gender

*   1
*   0

Converting categorical data into numerical form:


*   Makes data understandable for ML algorithms, Improves model training.

*   Allows mathematical operations on features.



